tests/ts/examples.ts

```diff
@@ -285,8 +285,8 @@ qb = qb.where(Person.raw('random()', 1, '2', raw('3')));
 qb = qb.alias('someAlias');
 
 // query builder hooks
-qb = qb.runBefore(async (result: any, builder: objection.QueryBuilder<Person>) => {});
-qb = qb.runAfter(async (result: Person[], builder: objection.QueryBuilder<Person>) => {});
+qb = qb.runBefore(async (result: any, builder: objection.QueryBuilder<Person, Person[]>) => {});
+qb = qb.runAfter(async (result: Person[], builder: objection.QueryBuilder<Person, Person[]>) => {});
```

Comment:
One should return stuff from thses. Whatever you return from `runBefore` gets passed to the next `runBefore` function as the first argument. That's impossible write typings for, but return value should be required.

Whatever you return from `runAfter` ends up being the return value of the query (in addition to being the input for the next `runAfter`). This we probably could write typings for?

Reply:
I wondered about that. I'll see what I can do.

Reply:
Okay, it seems that the purpose of these methods is to allow code to do an out-of-order prepending or appending of query actions to the query chain.

If the query chain were written expecting `runBefore` to change the result type that the first query method acts on, the type for the first method in the chain would be wrong, and the dev would have to typecast it. The typings only work if the final output of the `runBefore` series has the same type as the input to the first query method, which is now `Model | Model[]`.

I'm unable to find a way to incorporate `runAfter` result types. It would require adding a fourth type parameter to QueryBuilder and passing it from query builder to query builder through the typings. Call this type `AV`. We'd want `execute` to yield type `AV` when defined and `RV` otherwise, but AFAIK, Typescript cannot conditionally select among types. We could initialize `AV` to `never` so that `RV | never` evaluates to `RV`, but `RV | AV` would never evaluate to just `AV` because `RV` would never be `never`.

Unless there's a Typescript trick for doing this that I'm unaware of.